iceoryx_hoofs/source/posix_wrapper/periodic_timer.cpp

```diff
@@ -0,0 +1,172 @@
+
+using namespace iox::units::duration_literals;
+
+namespace iox
+{
+namespace posix
+{
+PeriodicTimer::PeriodicTimer(const iox::units::Duration interval) noexcept
+    : m_interval(interval)
+{
+    auto timerSemaphore = posix::Semaphore::create(posix::CreateUnnamedSharedMemorySemaphore, 0U);
+    cxx::Ensures(!timerSemaphore.has_error() && "Could not create Semaphore for PeriodicTimer!");
+    m_waitSemaphore = std::move(timerSemaphore.value());
+    start();
+}
+
+void PeriodicTimer::start() noexcept
+{
+    stop();
+    auto waitResult = m_waitSemaphore.timedWait(m_interval);
+    cxx::Ensures(!waitResult.has_error());
+    m_timeForNextActivation = now().value() + m_interval;
+}
+
+void PeriodicTimer::start(const iox::units::Duration interval) noexcept
+{
+    m_interval = interval;
+    start();
+}
+
+void PeriodicTimer::stop() noexcept
+{
+    if (*(m_waitSemaphore.getValue()) == static_cast<int>(posix::SemaphoreWaitState::TIMEOUT))
+    {
+        auto stopResult = m_waitSemaphore.post();
+        cxx::Ensures(!stopResult.has_error());
+    }
+}
+
+cxx::expected<units::Duration, TimerErrorCause> PeriodicTimer::now() noexcept
+{
+    struct timespec ts;
+    auto result =
+        posixCall(clock_gettime)(CLOCK_REALTIME, &ts).failureReturnValue(-1).evaluate();
+
+    if (result.has_error())
+    {
+        return createErrorCodeFromErrNo(result.get_error().errnum);
+    }
+
+    return cxx::success<units::Duration>(ts);
+}
+
+cxx::expected<iox::posix::WaitResult, TimerErrorCause> PeriodicTimer::wait(TimerCatchupPolicy policy) noexcept
+{
+    // To check if the TIMER is active (if the sempahore is acquired)
+    if (*(m_waitSemaphore.getValue()) == static_cast<int>(posix::SemaphoreWaitState::TIMEOUT))
+    {
+        if (now().value() > m_timeForNextActivation)
+        {
+            if (policy == TimerCatchupPolicy::IMMEDIATE_TICK)
+            {
+                m_timeForNextActivation = now().value();
+                m_waitResult.state = iox::posix::TimerState::TICK;
+                return cxx::success<iox::posix::WaitResult>(m_waitResult);
+            }
+            else if (policy == TimerCatchupPolicy::SKIP_TO_NEXT_TICK)
+            {
+                auto delay = now().value() - m_timeForNextActivation; // calculate the time delay
+                if (delay > m_interval)
+                {
+                    auto totalSlotToBeSkipped =
+                        delay.toMilliseconds() / m_interval.toMilliseconds(); // calculate the total slots missed
+                    m_timeForNextActivation =
+                        m_timeForNextActivation + m_interval * totalSlotToBeSkipped; // Skip to the next activation
+                }
+                else
+                {
+                    m_timeForNextActivation = m_timeForNextActivation + m_interval; // Skip to the next activation
+                }
+                auto timeDiff = m_timeForNextActivation - now().value(); // calculate remaining time for activation
```

Comment:
Accessing expected value without checking if an error is contained.

Reply:
Yes. modified to check if the expected has errors.